### PR TITLE
ISPN-11443 PersistenceManagerImpl thread checks need to be updated

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-11443
https://issues.redhat.com/browse/ISPN-11450
https://issues.redhat.com/browse/ISPN-11295

### CODE REVIEW

File: core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java [hunk 1]
```diff
@@ -29,6 +30,8 @@ public void applyTo(BlockHound.Builder builder) {
       // Block designates methods that should only hold a lock very briefly
       {
          CommonsBlockHoundIntegration.allowPublicMethodsToBlock(builder, OffHeapConcurrentMap.class);
+         // This acquires the locks from the above OffHeapConcurrentMap instance
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java]: Not sure what `above` means here, but the `lruLock` is in `SegmentedBoundedOffHeapDataContainer` itself.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java [hunk 2]
```diff
@@ -55,19 +56,22 @@
...
    private TimeService timeService;
    @SuppressWarnings("CanBeFinal")
    private volatile LockPlaceHolder current;
 
    /**
     * Creates a new instance.
     *
-    * @param executor
+    * @param blockingExecutor
+    * @param nonBlockingExecutor
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: I suggest removing them if they don't need a description

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java [hunk 3]
```diff
@@ -412,13 +418,13 @@ public InvocationStage toInvocationStage(Supplier<TimeoutException> timeoutSuppl
...
          }
-         return new SimpleAsyncInvocationStage(notifier.thenApplyAsync(state -> {
+         return new SimpleAsyncInvocationStage(CompletionStages.continueOnExecutor(notifier.thenApplyAsync(state -> {
             Object rv = checkState(state, () -> null, throwable -> throwable, timeoutSupplier);
             if (rv != null) {
                throw (RuntimeException) rv;
             }
             return null;
-         }, executor));
+         }, blockingExecutor), nonBlockingExecutor, this));
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: This should be in a separate commit for ISPN-11295.

Also, I would expect no blocking in the vast majority of cases, so we need an optimized path for that case and/or a PerfAck test to see if the extra thread switching affects performance.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java [hunk 4]
```diff
@@ -26,16 +26,19 @@
...
 
    public PerKeyLockContainer() {
       lockMap = new ConcurrentHashMap<>(INITIAL_CAPACITY);
    }
 
    @Inject
-   void inject(@ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor executor, TimeService timeService) {
-      this.executor = executor;
+   void inject(@ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor,
+         @ComponentName(KnownComponentNames.NON_BLOCKING_EXECUTOR) Executor nonBlockingExecutor, TimeService timeService) {
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java]: I'd use field injection instead.

`lock.setTimeService(timeService)` looks like it might be useful, but because there is no synchronization there's always a chance of missing one of the locks, so I'd remove that and force tests to replace the time service ahead of time.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java [hunk 5]
```diff
@@ -34,10 +34,11 @@ public StripedLockContainer(int concurrencyLevel) {
    }
 
    @Inject
-   void inject(@ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor executor, TimeService timeService) {
+   void inject(@ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor blockingExecutor,
+         @ComponentName(KnownComponentNames.NON_BLOCKING_EXECUTOR) Executor nonBlockingExecutor, TimeService timeService) {
       for (int i = 0; i < sharedLocks.length; i++) {
          if (sharedLocks[i] == null) {
-            sharedLocks[i] = new InfinispanLock(executor, timeService);
+            sharedLocks[i] = new InfinispanLock(blockingExecutor, nonBlockingExecutor, timeService);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java]: We should set the executors on the other branch as well. Even though I'd rather avoid that other branch completely, there's too much code in the tests that relies on replacing components after startup.

File: core/src/test/java/org/infinispan/lock/InfinispanLockTest.java [hunk 6]
```diff
@@ -41,7 +41,7 @@ public void testTimeout() throws InterruptedException {
       final String lockOwner1 = "LO1";
       final String lockOwner2 = "LO2";
 
-      final InfinispanLock lock = new InfinispanLock(commonPool(), AbstractCacheTest.TIME_SERVICE);
+      final InfinispanLock lock = new InfinispanLock(commonPool(), commonPool(), AbstractCacheTest.TIME_SERVICE);
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/lock/InfinispanLockTest.java]: Please use `AbstractInfinispanTest.testExecutor()` in the tests, with the common pool it's impossible to filter logs by test.

File: core/src/test/java/org/infinispan/lock/LockContainerTest.java [hunk 7]
```diff
@@ -32,29 +30,28 @@
  */
 @Test(groups = "unit", testName = "lock.LockContainerTest")
 public class LockContainerTest extends AbstractInfinispanTest {
-   private final ExecutorService executor = new WithinThreadExecutor();
 
    public void testSingleLockWithPerEntry() throws InterruptedException {
       PerKeyLockContainer lockContainer = new PerKeyLockContainer();
-      TestingUtil.inject(lockContainer, executor, AbstractCacheTest.TIME_SERVICE);
+      TestingUtil.inject(lockContainer, AbstractCacheTest.TIME_SERVICE);
```
Comment (danberindei) [inline core/src/test/java/org/infinispan/lock/LockContainerTest.java]: I'm a bit scared that none of these tests need the executor, it suggests that maybe they don't cover all the code paths ;)

File: core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/CoreBlockHoundIntegration.java]: The issue that identified it wasn't actually `lruLock`, but that it found it was the stamped lock from the off heap concurrent map.  However this fixes both :) I can change the comment.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: In vast majority of cases this code is never even hit, because the notifier is normally usually done above. So I would think the thread switching here is probably not going to be an issue, and actually has a very high likely hood of actually blocking.

The reason the change is needed is because it then resumed the operation on the blocking thread, which we don't want.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java (hunk 4 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java]: > I'd use field injection instead.

I can change it, was just keeping it consistent with the previous code.

> `lock.setTimeService(timeService)` looks like it might be useful, but because there is no synchronization there's always a chance of missing one of the locks, so I'd remove that and force tests to replace the time service ahead of time.

I thought you said before that cache initialization was thread safe in regards to visibility? Is this not true now?

File: core/src/test/java/org/infinispan/lock/InfinispanLockTest.java (hunk 6 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/lock/InfinispanLockTest.java]: Sure, I can change the existing reference to the updated one.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java (hunk 5 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java]: I think I will just remove the if/else block.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: I can make a separate commit for it.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: I will add a description.

File: core/src/test/java/org/infinispan/lock/LockContainerTest.java (hunk 7 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/lock/LockContainerTest.java]: I agree. As I mentioned above the executor is only used when it has to wait for the lock.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java (hunk 3 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: I will take a closer look at this in general with this pass though. I should be able to solve it how I would want for ISPN-11295.

**Comment by wburns** [comment]
@danberindei updated !

**Comment by wburns** [comment]
The change for the lock was simpler than expected, I just resume it on the non blocking thread to make sure that the caller who releases the lock doesn't get stuck. Not quite sure why I had found this required blocking pool before. Guessing I resolved a different issue elsewhere.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java (hunk 4 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/PerKeyLockContainer.java]: Regular cache initialization is safe. But once a component A is started, there's no other synchronization to ensure that changes made to A's fields are properly seen by all threads that are concurrently accessing A.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java (hunk 5 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java]: I'm not sure that's ok either, as it will basically drop the locks that are held by in-progress operations/txs, and the dropped locks will never resume their waiters.

And unfortunately modules that need to override core components use the same `rewire()` functionality as tests, so we can't say that it will only affect tests (although modules should do this before any locks can be acquired, there's no guarantee).

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: > In vast majority of cases this code is never even hit, because the notifier is normally usually done above. 

That depends on the access access pattern. PerfAck tests make each stressor use separate keys, so there is no contention, but when there is lots of contention the stage wouldn't be done.

The new approach resuming directly on a non-blocking thread looks much better!

Comment (wburns) [reply core/src/main/java/org/infinispan/util/concurrent/locks/impl/InfinispanLock.java]: Yeah, I am not sure why it needed a blocking thread before. So I am glad to change it to this now.

File: core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java [hunk 8]
```diff
@@ -34,13 +34,10 @@ public StripedLockContainer(int concurrencyLevel) {
...
-   void inject(@ComponentName(KnownComponentNames.BLOCKING_EXECUTOR) Executor executor, TimeService timeService) {
+   void inject(@ComponentName(KnownComponentNames.NON_BLOCKING_EXECUTOR) Executor nonBlockingExecutor,
+         TimeService timeService) {
       for (int i = 0; i < sharedLocks.length; i++) {
-         if (sharedLocks[i] == null) {
-            sharedLocks[i] = new InfinispanLock(executor, timeService);
-         } else {
-            sharedLocks[i].setTimeService(timeService);
-         }
+         sharedLocks[i] = new InfinispanLock(nonBlockingExecutor, timeService);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/util/concurrent/locks/impl/StripedLockContainer.java]: I'm starting to think figuring out why a lock isn't unlocked after a `BasicComponentRegistry.rewire()` is too hard, and tests almost never replace executors anyway, so the master version of the method is safer.

**Comment by wburns** [comment]
Updated

**Comment by wburns** [comment]
Trying a different approach for commits in preload. Will see what CI says.

**Comment by wburns** [comment]
The failure https://ci.infinispan.org/job/Infinispan/job/PR-8022/4/testReport/junit/org.infinispan.statetransfer/StateTransferDistSharedCacheLoaderFunctionalTest/testSharedFetchedStoreEntriesUnaffected/ is related. Looking.

**Comment by wburns** [comment]
Should be fixed. I will also log a JIRA for https://ci.infinispan.org/job/Infinispan/job/PR-8022/4/testReport/junit/org.infinispan.remoting.transport/InitialClusterSizeTest/Blocking/ which is unrelated to this PR.

**Comment by wburns** [comment]
> Should be fixed. I will also log a JIRA for https://ci.infinispan.org/job/Infinispan/job/PR-8022/4/testReport/junit/org.infinispan.remoting.transport/InitialClusterSizeTest/Blocking/ which is unrelated to this PR.

Is actually fixed in https://github.com/infinispan/infinispan/pull/7997

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java [hunk 9]
```diff
@@ -1384,25 +1388,42 @@ private long getMaxEntries() {
...
-         //commitIfNeeded can throw an exception, so we need a try { } finally { }
-         resumeIfNeeded(transaction);
+         stage = (CompletionStage) putStage.whenCompleteAsync((pendingTransaction, t) -> {
+            try {
+               transactionManager.resume(pendingTransaction);
+               commitIfNeeded(t == null);
+            } catch (InvalidTransactionException | SystemException e) {
+               throw new PersistenceException("Unable to preload!", e);
+            } finally {
+               //commitIfNeeded can throw an exception, so we need a try { } finally { }
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: Not sure the comment belongs here

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java [hunk 10]
```diff
@@ -1384,25 +1388,42 @@ private long getMaxEntries() {
...
    }
 
-   private void preloadKey(AdvancedCache<Object, Object> cache, MarshallableEntry me) {
-      final Transaction transaction = suspendIfNeeded();
-      boolean success = false;
-      try {
+   private Single<?> preloadKey(AdvancedCache<Object, Object> cache, MarshallableEntry me) {
+      // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
+      InternalMetadataImpl metadata = new InternalMetadataImpl(me.getMetadata(), me.created(), me.lastUsed());
+      CompletionStage<Object> stage;
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]:
```suggestion
      CompletionStage<?> stage;
```

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java [hunk 11]
```diff
@@ -1384,25 +1388,42 @@ private long getMaxEntries() {
...
+            } finally {
+               //commitIfNeeded can throw an exception, so we need a try { } finally { }
+               resumeIfNeeded(transaction);
+            }
+         }, blockingExecutor);
+      } else {
+         stage = cache.putAsync(me.getKey(), me.getValue(), metadata);
       }
+      return RxJavaInterop.completionStageToMaybe(stage)
+            .toSingle(me);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: The return value isn't used, so `completionStageToSingle()` should work -- or there should be a comment explaining why the double conversion is needed.

Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: It is used sadly because stage can have a null value and we have to have a non null one. Same reason why I can't use `CompletionStage<?>` above.

File: core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java (hunk 9 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java]: Sure I can remove.

**Comment by wburns** [comment]
Updated and rebased. No code changes.

**Comment by danberindei** [comment]
Merged, thanks Will!
